File: control_center/keypad/process.py
import sys
import time
from Phidget22.Phidget import *
from Phidget22.Devices.DigitalOutput import *
from Phidget22.Devices.DigitalInput import *


number_of_inputs = 4
number_of_outputs = 4
timeout_duration = 15 * 1000 # its in miliseconds

# create input pins
input_pins  = [ DigitalInput() for each_index in range(number_of_inputs)  ]
for each_index, each in enumerate(input_pins):
    each.setChannel(each_index)
    each.openWaitForAttachment(timeout_duration)

# create ouput pins
output_pins = [ DigitalOutput() for each_index in range(number_of_outputs) ]
for each_index, each in enumerate(output_pins):
    each.setChannel(each_index)
    each.openWaitForAttachment(timeout_duration)

def which_key(input_pin, index_of_input):
    key_values = [
        ["1", "2", "3", "A"],
        ["4", "5", "6", "B"],
        ["7", "8", "9", "C"],
        ["*", "0", "#", "D"]
    ]
    return_value = None
    
    # do trial-and-error to find which output pin
    for index_of_output, each_output in enumerate(output_pins):
        each_output.setState(1)
        time.sleep(0.05)
        if input_pin.getState() == 0:
            return_value = key_values[index_of_input][-(index_of_output+1)]
            break
    
    # reset all the output pins
    for each in output_pins:
        each.setState(0)
    
    time.sleep(0.05)
    
    return return_value


previous_vals = [0, 0, 0, 0]
current_values = [0, 0, 0, 0]
while True:
    # find which one(s) changed 
    for index_of_input, (old_value, input_pin) in enumerate(zip(previous_vals, input_pins)):
        # update the input pin value right before checking it
        value = current_values[index_of_input] = input_pin.getState()
        if value == 1 and value != old_value:
            print(which_key(input_pin, index_of_input))
            sys.stdout.flush()
    
    # keep track of old states
    previous_vals = current_values

# clean up
for each in input_pins + output_pins:
    each.close()

File: control_center/keypad/process.js
let phidget22 = require("phidget22")
// module.exports = {
//     setupListener: whenStarPressed
// }


// 
// fixed data
// 
const numberOfInputs = 4
const numberOfOutputs = 4
const timeoutDuration = 15 * 1000 // its in miliseconds
const serialNumber = 96781
const probablyAPortNumber = 5661

// 
// helper for setting up pins
// 
let onStateChange
let allPins = []
let openPromiseList = []
let conn = new phidget22.Connection(probablyAPortNumber, "localhost")
let createPins = (numberOfPins, inputOrOutput)=> {
    let pins = []
    
    for (let eachIndex in [...Array(numberOfInputs)]) {
        // setup
        let newPin = new phidget22[inputOrOutput]()
        if (serialNumber) {
            newPin.setDeviceSerialNumber(serialNumber)
        }
        newPin.setChannel(eachIndex)
        
        // connect all inputs to a single state-change function
        if (inputOrOutput == "DigitalInput") {
            newPin.onStateChange = (...args)=> onStateChange(newPin, eachIndex, ...args)
        }
        // log onAttach/Detach
        newPin.onAttach = ()=>console.log(`${inputOrOutput} pin ${eachIndex} was attached`)
        newPin.onDetach = ()=>console.log(`${inputOrOutput} pin ${eachIndex} was attached`)
        
        // try to open the pins
        let pinLoaded = newPin.open(timeoutDuration)
        openPromiseList.push(pinLoaded)

        // add to output array
        pins.push(newPin)
        // add to all-pin array
        allPins.push(newPin)
    }

    return pins
}
let closeAllPins = ()=> {
    for (let each of allPins) {
        each.close()
    }
}

// 
// helper for detecting which key
//
let wait = (seconds)=>new Promise(resolve=>setTimeout(_=>resolve(),seconds*1000))
let whichKey = async (inputPin, indexOfInput, outputPins) => {
    let setStateDelayTime = 0.05
    keyValues = [
        ["1", "2", "3", "A"],
        ["4", "5", "6", "B"],
        ["7", "8", "9", "C"],
        ["*", "0", "#", "D"]
    ]
    returnValue = null
    
    
    for (let eachOutputIndex in outputPins) {
        outputPins[eachOutputIndex].setState(1)
        await wait(setStateDelayTime)
        if (inputPin.getState() == 0) {
            returnValue = keyValues[indexOfInput][numberOfOutputs-(eachOutputIndex+1)]
        }
    }
    
    for (let eachOutput of outputPins) {
        eachOutput.setState(0)
    }
    await wait(setStateDelayTime)
    
    return returnValue
}


// 
// main function
// 
conn.connect().then(() => {

    // create pins
    outputPins = createPins(numberOfOutputs, "DigitalOutput")
    inputPins = createPins(numberOfOutputs, "DigitalInput")

    // what to do when an input value changes
    onStateChange = async (pin, pinIndex, state) => {
        try {
            console.log(`keypad state changed: pin:${pinIndex} state: ${state}`)
            if (state == true) {
                
            }
        } catch (error) {
            console.log(`error is:`,error)
        }
    }
    
    // what do do after all pins have connected (or failed)
    Promise.all(openPromiseList).then(()=> {
        // NOTE: un-commenting this crashes the code for some reason, even though its similar to the online example
        // for (let each of outputPins) {
        //     each.setDutyCycle(1).catch((err)=>{
        //         console.error('Error during setDutyCycle:', err)
        //     })
        // }
        console.log(`all pins opened`)
        // closeAllPins()
    })
}).catch(error=> console.error(error))


